### PR TITLE
disallow full table scans

### PR DESCRIPTION
**Pull request by davepacheco**
tl;dr: RFD 192 talks at length about why we want to avoid table scans.  Ideally, we'd like to know at build-time if we write queries that will necessitate table scans, but this is very hard.  One easy thing we _can_ do is enable CockroachDB's `disallow_full_table_scans` option, which causes it to fail queries that would necessitate a table scan.  If the test suite passes, and we believe we have good test suite coverage for Nexus, then this gives us decent confidence that we haven't done something grossly wrong in this area.  See #254 for more discussion on this problem and different approaches for solving it.

---

In #254, [I tried](https://github.com/oxidecomputer/omicron/issues/254#issuecomment-926243830) using this option before and found way too many false positives to be useful.  In [v21.2.2](https://www.cockroachlabs.com/docs/releases/v21.2#v21-2-2) (which we've now updated past), CockroachDB has made this update:

> The optimizer has been updated so that if [disallow_full_table_scans](https://www.cockroachlabs.com/docs/v21.2/set-vars) is true, it will never plan a full table scan with an estimated row count greater than large_full_scan_rows. If no alternative plan is possible, an error will be returned, just as it was before. However, cases where an alternative plan is possible will no longer produce an error, since the alternative plan will be chosen. As a result, users should see fewer errors due to disallow_full_table_scans.

Now that we've updated past that, I thought I'd resurrect that experiment to see if we can automatically fail queries that would attempt a table scan before they do so.  It worked much better this time!

There were three test failures:

- The db pagination tests create their own table and _do_ full scans against it.  That's okay.  I just added indexes to appease the new check.
- The silo delete code path soft-deletes all silo users for that silo, but there's no corresponding index on silo_user.  Additionally, the query here was missing a `time_deleted IS NULL` filter.
- The updates code path soft-deletes outdated artifacts but was missing an index to do this.

In other words, it caught 2-3 real bugs!  The error messages were pretty good, too:

>[2022-04-29T22:05:56.730516441Z]  INFO: 2bdbae9a-3002-48fa-8b91-6285bfb28947/dropshot_external/17282 on ivanova: request completed (req_id=4d175a9c-dfe7-4f99-a0bd-df6c08054f8c, uri=/silos/discoverable, method=DELETE, remote_addr=127.0.0.1:38953, local_addr=127.0.0.1:39314, error_message_external="Internal Server Error", response_code=500)
>    error_message_internal: database error (kind = Unknown): query `UPDATE "silo_user" SET "time_deleted" = $1 WHERE ("silo_user"."silo_id" = $2)` contains a full table/index scan which is explicitly disallowed
>    HINT: try overriding the `disallow_full_table_scans` or increasing the `large_full_scan_rows` cluster/session settings

>[2022-04-29T22:06:00.51476957Z]  INFO: 3ac214f0-033f-4e29-b2f5-4ddbd8f662aa/dropshot_external/17282 on ivanova: request completed (req_id=8ec3a7ec-30ef-40cc-9cea-89134027b3d9, uri=/updates/refresh, method=POST, remote_addr=127.0.0.1:60077, local_addr=127.0.0.1:35522, error_message_external="Internal Server Error", response_code=500)
>    error_message_internal: error deleting outdated available artifacts: Connection(Query(DatabaseError(Unknown, "query `DELETE FROM \"update_available_artifact\" WHERE (\"update_available_artifact\".\"targets_role_version\" < $1)` contains a full table/index scan which is explicitly disallowed")))

(That's all part of the "internal_message" of the 500 error that gets logged with the request.)

So I think this is a good change.

There are two risks here:

- We find false positives and it gets in the way of future development.  I'm optimistic that won't be a problem just because of how few false positives we had after never having used this before.
- We find some code path that's not tested by the test suite that we would rather have function in production even if it meant a table scan.  We can't know ahead of time what those code paths are -- if we did, we'd cover them in the test suite.  It'd be nice to have an escape hatch somehow, but that'd be risky too.  And without a way to tweak configuration at runtime, it doesn't seem useful to do that right now.

### CODE REVIEW

**Review by david-crespo — APPROVED** [review]
Very nice! My worry is that `large_full_scan_rows` is too high and our test tables don’t have enough rows to trigger it. What is it by default, and should we set it low on purpose for the tests?

Edit: The default value is 1000.

**Comment by davepacheco** [comment]
> Very nice! My worry is that `large_full_scan_rows` is too high and our test tables don’t have enough rows to trigger it. What is it by default, and should we set it low on purpose for the tests?

Yeah, I wondered about this too and I don't fully understand.  If I use `omicron-dev db-run`, connect, and check `large_full_scan_rows`, it's 1000.  Nonetheless, we tripped the check in the three cases above (none of which has much in the table), _plus_ the datastore test that I added (which explicitly verifies the check trips on an empty table).  If we're wrong about this, the risk _is_ unfortunate: it means things would probably work in the test suite and we'd start seeing errors all over the place as the system gets larger.  But I think for that to be true, at the very least, the test I added to datastore.rs should catch it.

What do you think?

**Comment by davepacheco** [comment]
Okay, I was writing up something to ask the folks at CockroachDB and I think I see what's going on.  tl;dr: I think we _do_ want to try setting `large_full_scan_rows = 0` as well.

I think what's going on is that when you first create a table, CockroachDB doesn't have any statistics about it.  When you run a query for which it needs to a table scan and it has no statistics, it seems to assume the worst -- that it's larger than `large_full_scan_rows` (or at least larger than the default value, 1000).  A few seconds later, it will analyze the table and collect statistics.  At that point it thinks the table's much smaller and will _not_ fail this check.  You can trigger the "analyze" by hand, too.  Check it out.  Here's the setup:

```
root@127.0.0.1:32221/omicron> show large_full_scan_rows;
  large_full_scan_rows
------------------------
  1000
(1 row)


Time: 1ms total (execution 1ms / network 0ms)

root@127.0.0.1:32221/omicron> show disallow_full_table_scans;
  disallow_full_table_scans
-----------------------------
  on
(1 row)


Time: 1ms total (execution 1ms / network 0ms)

root@127.0.0.1:32221/omicron> create table foo (s int);
CREATE TABLE


Time: 5ms total (execution 5ms / network 0ms)

```

If I try immediately to select the whole table, it fails:

```
root@127.0.0.1:32221/omicron> select * from foo;
ERROR: query `select * from foo` contains a full table/index scan which is explicitly disallowed
SQLSTATE: P0003
HINT: try overriding the `disallow_full_table_scans` or increasing the `large_full_scan_rows` cluster/session settings
```

and the `EXPLAIN` shows that it has no stats:

```
root@127.0.0.1:32221/omicron> EXPLAIN select * from foo;
          info
------------------------
  distribution: full
  vectorized: true

  • scan
    missing stats
    table: foo@primary
    spans: FULL SCAN
(7 rows)


Time: 1ms total (execution 1ms / network 0ms)

```

Now I can explicitly analyze it and run EXPLAIN again:

```
root@127.0.0.1:32221/omicron> analyze foo;
ANALYZE


Time: 14ms total (execution 14ms / network 0ms)

root@127.0.0.1:32221/omicron> EXPLAIN select * from foo;
                                     info
-------------------------------------------------------------------------------
  distribution: full
  vectorized: true

  • scan
    estimated row count: 1 (100% of the table; stats collected 2 seconds ago)
    table: foo@primary
    spans: FULL SCAN
(7 rows)


Time: 1ms total (execution 1ms / network 0ms)

```

and now it's happy to select the whole table because it thinks there's only one row, which is way less than the limit:

```
root@127.0.0.1:32221/omicron> select * from foo;
  s
-----
(0 rows)


Time: 1ms total (execution 1ms / network 0ms)

```

I found all this because I was putting an example together to ask the CockroachDB folks and discovered that after a few minutes we no longer were hitting the check.

**Comment by davepacheco** [comment]
The good news is that with large_full_scan_rows = 0 in place, the test suite still passes for me.
